### PR TITLE
Generate SDSC for symbolic batch dimension for pointwise ops.

### PR DESCRIPTION
**Pull request by nethra-khandige**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:
Adds initial support for symbolic (dynamic) batch size (dim0) dimension in the SDSC codegen pipeline for pointwise ops.

Bounds propagation (`pass_utils.py`, `op_spec.py`, `spyre_kernel.py`):

- Adds `compute_symbolic_bounds` to extract (max, granularity) from ShapeEnv at codegen time
- Adds `symbolic_dim_bounds`: dict[str, tuple[int, int]] to OpSpec so bounds survive when ShapeEnv is no longer active.

SDSC codegen (`superdsc.py`, `compute_ops.py`):

- Adds `_resolve_sdsc_size` to use max from `symbolic_dim_bounds` (not size_hint) when building the SDSC iteration space
- Adds symbolic_dims field to SDSCSpec carrying sdsc dim name → (pytorch_sym_name, max, granularity)
- Adds `SymbolKind.dimension `variant for dynamic dim symbols registered in the SDSC bundle symbol table

Views (`views.py`):

- Preserves symbolic range expressions through align_tensors so downstream passes (work division, SDSC codegen) see the actual symbolic sizes rather than concretised hints.

Additional logs/sdsc.json: 
[symbolic_dim.zip](https://github.com/user-attachments/files/28930890/symbolic_dim.zip)
(contains for symbolic dim and also for symbolic dim + symbolic address)

#### Which issue(s) this PR is related to:
#43, #2288, #220 

#### Does this PR introduce a user-facing change?
Yes. Users can now use mark_dynamic(dim, min=..., max=...) on a batch dim for pointwise ops to generate SDSC.json.

#### Additional note:
To be merged after #2499

### CODE REVIEW

**Comment by nethra-khandige** [comment]
@dgrove-oss @tardieu  @pradghos

This PR enables SDSC generation for symbolic batch size dimension.
Example
Test file:
```

import torch
import torch._dynamo as dynamo
import torch.nn.functional as F

DEVICE = torch.device("spyre")
torch.manual_seed(0xAFFE)

def gelu_fn(a):
    return F.gelu(a)

x = torch.rand(560, 1024, dtype=torch.float16)

# Mark dim 0 as dynamic with an explicit upper bound.
x_device = x.to(DEVICE)
dynamo.mark_dynamic(x_device, 0, min=56, max=616)
compiled_fn = torch.compile(gelu_fn)

compiled_result = compiled_fn(x_device).cpu()
print(f"Spyre Compiled result\n{compiled_result}")

```

SDSC: (generated as per golden SDSC provided by the DeepTools team)
[symbolic_sdsc.json](https://github.com/user-attachments/files/29099830/symbolic_sdsc.json)

Detailed logs:
[symbolic_sdsc.log](https://github.com/user-attachments/files/29099843/symbolic_sdsc.log)

cc: @Vivek1258 @joyalbin

**Comment by dgrove-oss** [comment]
Initial review from coding agent.  Please take a look at the suggestions in the summary and either address or add a comment saying they should be deferred to a future PR.


-----

## PR #2673 — Generate SDSC for symbolic batch dimension for pointwise ops

**Author:** Nethra Khandige | **+355 / -22** across 8 files | **CI:** All green

---

### Overview

This PR wires dynamic batch-size (dim0) support through the full SDSC codegen pipeline for pointwise ops. The key mechanism: at `create_op_spec` time (when `ShapeEnv` is still live), it extracts `(max, granularity)` bounds and stores them in `OpSpec.symbolic_dim_bounds`. Downstream SDSC generation then uses those serialized ints—avoiding a dependency on the live `ShapeEnv` during codegen, which is already gone at that stage. Additionally `views.py`/`align_tensors` is patched to propagate symbolic expressions forward instead of always concretizing them.

---

### Code Quality

**Strengths:**
- The separation of concerns is clean: bounds extraction (`pass_utils.py`) stays independent of codegen, and the `ShapeEnv`-independence is architecturally sound.
- `_per_core_symbolic_dim_info` being factored out to keep `ss_` and `el_` blocks byte-for-byte identical is a good catch.
- The `_bounded_or_hint` helper in `views.py` is a principled guard against propagating unbounded auto-dynamic symbols.
- Unit tests are thorough: `TestResolveSdscSize` and `TestSymbolKindDimension` cover concrete/symbolic paths and sentinel fields.

**Issues to address:**

1. **`_resolve_sdsc_size` only handles single-free-symbol expressions** (`next(iter(expr.free_symbols))`). For a multi-symbol expression this silently falls back to `_concretize_for_sdsc` (a size_hint). That's probably fine today since only single-symbol batch dims are supported, but the silent fallback is dangerous. Consider either asserting `len(expr.free_symbols) == 1` when bounds is found, or at minimum adding a comment explaining the intent.

2. **`_bounded_or_hint` has a double-application problem in `views.py`**: the function is called once in the main unsplit-dim branch and then *again* in the post-loop restore pass. If `orig_ranges[var] == var_ranges[var]` (concrete dim or already concretized), the second call is redundant. Worth verifying the two calls are not interacting unexpectedly for dimensions that get concretized to the same size as their hint.

3. **Private import leaking into public API** (`spyre_kernel.py`): `_finite_upper_or_none` is imported from `pass_utils` with its private underscore name. Either export it properly (remove the `_` prefix in `pass_utils.py`) or add a comment explaining why the private function is intentionally imported here. The same name is also accessed via a local import in `views.py` — there's inconsistency in how this helper is accessed.

4. **PR description checkbox formatting**: The PR template has `- [  ] feature` (double space inside brackets) — this renders as unchecked. Minor, but worth fixing.

---

### Specific Code Observations

**`compute_ops.py` — symbol ID layout:**

The ID assignment and layout comment are helpful. However, `dim_local_symbols` deduplicates by `pytorch_sym`, while the iteration is over `symbolic_dims` which is keyed by SDSC dim name — these could diverge if two SDSC dims map to the same PyTorch symbol. The dedup check `if pytorch_sym not in dim_local_symbols` is correct but might be surprising; worth a comment explaining that multiple SDSC dims can share a PyTorch symbol.

**`superdsc.py` — `parse_op_spec` string-keyed bounds lookup:**

```python
sym_str = str(size_expr)
if sym_str in op_spec.symbolic_dim_bounds:
```

Using `str(size_expr)` as a dict key is fragile — sympy's `__str__` is not guaranteed to be canonical across versions. The round-trip is consistent (both sides use `str(size_expr)`), but a comment noting this consistency contract would help future readers.

**`superdsc.py` — truncated comment:**

```python
# For symbolic dims, use the max from symbolic_dim_bounds as the iteration-space size
# so the emitted SDSC JSON is generated with max
```

"generated with max" — with max what? The sentence is cut off.

**`pass_utils.py` — `compute_symbolic_bounds` redundant guard:**

```python
if isinstance(expr, (int, sympy.Integer)):
    return None
if not (hasattr(expr, "free_symbols") and expr.free_symbols):
    return None
```

The `isinstance` check already handles concrete ints; the `hasattr` guard is redundant for those. Fine as a defensive measure, but slightly confusing given the `Union[Expr, int]` type hint. A short comment would clarify the intent.

**`test_codegen.py` — `dynamic=False` change:**

```python
comp_fn = torch.compile(fn, dynamic=False)
```

`test_symbolic_batch_dim_pointwise_split` is changed from `dynamic=True` to `dynamic=False`. Using `dynamic=False` with `mark_dynamic` is the correct pattern (Dynamo defers to per-tensor marks), but there should be a comment explaining why `dynamic=True` no longer works here.

---

### Test Coverage

- Unit tests for `_resolve_sdsc_size` and `SymbolKind.dimension` are solid.
- `compute_symbolic_bounds` tests cover finite bounds, infinite upper, `None` shape env, and concrete inputs. Good.
- **Missing**: No test exercises the full end-to-end path verifying the emitted SDSC JSON contains correct `dimToSymbolMapping_`, `symbolicDimInfo_`, and `inputSymbolsAndTags_` fields. The existing integration test (`test_symbolic_batch_dim_pointwise_split`) only asserts on split values. Consider asserting on JSON structure for at least one simple case.
- **Missing**: `_per_core_symbolic_dim_info` has no dedicated unit tests — its division logic (`max_val // wk_slices`, `max(1, granularity // wk_slices)`) is easy to get wrong. A few cases (e.g. `wk_slices=1`, `wk_slices=32`, `wk_slices > granularity`) would give confidence.

---

### Potential Risks

- The `dynamic=False` change in `test_symbolic_batch_dim_pointwise_split` changes test semantics — verify this doesn't silently weaken coverage.
- The double restore in `views.py` is subtle; the condition `new_var_ranges[var] != var_ranges[var]` as a guard for "a split was applied" could be wrong if a split happened to produce the same numerical value as the original. Worth a comment or assertion.
- `SymbolKind` is a dataclass used across codegen — adding three new fields (`granularity`, `max_value`, `pytorch_sym`) with default sentinels means existing `kernel`/`pool`/`kernel_derived` instances silently carry those fields at `0`/`""`. If any downstream code ever iterates `SymbolKind` fields generically this could cause issues. Low risk today but worth noting.

---

### Summary

Solid architecture, good test coverage for the new utilities. The main things to address before merging:

1. Fix the truncated comment in `superdsc.py` ("generated with max").
2. Add unit tests for `_per_core_symbolic_dim_info`.
3. Explain the `dynamic=True` → `dynamic=False` change in `test_symbolic_batch_dim_pointwise_split`.
4. Clarify the `_finite_upper_or_none` import inconsistency (private name imported across modules).
5. Consider a smoke test asserting SDSC JSON structure contains the expected symbolic dim fields.

**Comment by nethra-khandige** [comment]
@dgrove-oss ,Thanks for the review. I have updated the PR accordingly. 

**1. Truncated comment in `superdsc.py`** - fixed at [superdsc.py:710](torch_spyre/_inductor/codegen/superdsc.py#L710). Now reads: *"so the emitted SDSC JSON is generated with max sizes baked in, not symbols."*

**2. Unit tests for `_per_core_symbolic_dim_info`** - added `TestPerCoreSymbolicDimInfo` in [tests/inductor/test_codegen.py](tests/inductor/test_codegen.py). 

**3. `dynamic=False` in `test_symbolic_batch_dim_pointwise_split`** - added an inline comment
`mark_dynamic` is sufficient on its own; `dynamic=True` was a leftover from earlier in development before the symbolic propagation in `views.py` was complete. Scope here is dim-0 symbolic only.

**4. `_finite_upper_or_none` naming inconsistency** - renamed to `finite_upper_or_none` (dropped the leading underscore) across `pass_utils.py`, `work_division.py`, `spyre_kernel.py`, and `views.py`, plus the doc-comment reference in `test_inductor_fx_passes.py`. 

**5. SDSC JSON smoke test** - added `TestSdscJsonSymbolicDimSmoke`. In [tests/inductor/test_codegen.py](tests/inductor/test_codegen.py)

One thing I didn't pull in: `inputSymbolsAndTags_` (top-level on the SDSC dict, separate code path). Happy to extend the same test to assert on it too if you'd prefer it covered in this PR, please let me know.


The following things are deferred to a follow-up PR (flagging here so they don't get lost):

- **`_resolve_sdsc_size` silent multi-symbol fallback.** Agree the silent `_concretize_for_sdsc` fallback is fragile. Not reachable today because currently we only emits single-symbol batch dims, but I'll add either an `assert len(expr.free_symbols) == 1 when bounds is found` or an inline comment when the next multi-symbol case lands.
- **Double-application of `_bounded_or_hint` in `views.py`.** The post-loop pass is a no-op when the expression is already concrete (so it's not interacting unexpectedly with the unsplit-dim branch for concrete dims that happen to equal their hint), but some the redundancy is there. Will tighten the guard in the follow-up.

**Review by dgrove-oss — DISMISSED** [review]
Final code review had 3 minor suggestions, but they are optional.

Please resolve the merge conflict, optionally address the suggestions, and then we can merge.

-----

PR #2673: Generate SDSC for symbolic batch dimension for pointwise ops

Summary

Adds end-to-end symbolic-dim support for pointwise ops: propagates symbolic iteration-space ranges through align_tensors, serializes (max, granularity) bounds into OpSpec, and emits dimToSymbolMapping_ / symbolicDimInfo_ / inputSymbolsAndTags_ in the SDSC JSON.

---
Prior Comments — Status

Item 1 — Truncated stale comment in superdsc.py
✅ Addressed. The five-line "wiring: this helper has no call sites yet" block is removed from compute_granularity (pass_utils.py:151–155), and the matching TODO in compute_max_size is also cleaned up.

Item 2 — Unit tests for _per_core_symbolic_dim_info
✅ Addressed. TestPerCoreSymbolicDimInfo adds five parameterized cases covering the zero-dim, single dim, cross-core, non-symbolic, and multi-dim scenarios.

Item 3 — Comment explaining dynamic=False in smoke test
✅ Addressed. test_symbolic_batch_dim_pointwise_split now carries an inline comment explaining why dynamic=False is intentional (ShapeEnv is seeded by OpSpec bounds, not mark_dynamic at test time).

Item 4 — Rename _finite_upper_or_none → finite_upper_or_none
✅ Addressed. Renamed in pass_utils.py, with call sites updated in work_division.py, spyre_kernel.py, and views.py. All imports updated.

Item 5 — SDSC JSON smoke test asserting new fields
✅ Addressed. TestSdscJsonSymbolicDimSmoke asserts dimToSymbolMapping_ and symbolicDimInfo_ are present and correctly structured in the generated JSON.

All five prior blockers/suggestions are resolved.

---
New Code Review

One thing caught my eye in the new _bounded_or_hint / symbolic-range restoration logic in views.py:

QUESTION — The docstring at line 748 has a typo: "are not symbolic dims )" — stray ) in the middle of a sentence and missing closing period. Minor, but worth fixing. (views.py:748)

QUESTION — In _resolve_sdsc_size (superdsc.py:529–533), when expr.free_symbols has more than one symbol, next(iter(...)) picks an arbitrary one. Is it guaranteed at this call site that symbolic iteration-space sizes are always single-symbol expressions? If so, a short assert or comment would make that invariant explicit.

SUGGESTION — The two-pass restoration loop at the bottom of align_tensors (lines 793–796) and the in-loop _bounded_or_hint call (lines 770–772) both attempt to restore symbolic ranges for the same unsplit vars. The comment on line 765 acknowledges the in-loop path; the bottom loop then catches any that were missed. This is correct but double-updates are possible when both paths fire for the same var. Not a bug (idempotent), but worth a comment noting the two paths are complementary and not redundant.

---
Verdict

APPROVE — all five prior comments are fully addressed, CI is green, all 11 commits are signed. The new questions above are minor and non-blocking.

**Review by dgrove-oss — APPROVED** [review]
lgtm
